src/hipscat/pixel_math/margin_bounding.py

```diff
@@ -1,255 +1,150 @@
+            to save compute time if accuracy is less important.
     Returns:
         :obj:`numpy.array` of boolean values corresponding to the ra and dec
             coordinates checked against whether a given point is within any of the
             provided polygons.
     """
-    polar, pole = pm.pixel_is_polar(order, pix)
-
-    if not polar:
-        raise ValueError("provided healpixel must be polar")
-
-    part_pix_res = hp.nside2resol(2**order)
-    marg_pix_res = hp.nside2resol(2**margin_order)
-
-    # get the approximate number of boundary samples to cover a highest_k pixel
-    # on the boundary of the main pixel
-    boundary_range = int((marg_pix_res / part_pix_res) * step)
-    pixel_boundaries = hp.vec2dir(hp.boundaries(2**order, pix, step=step, nest=True), lonlat=True)
-
-    # to optimize our code, we only want to take boundary samples from the part
-    # of the pixel that directly abuts the polar margin pixels.
-    if pole == "North":
-        end = len(pixel_boundaries[0])
-        east_ra = pixel_boundaries[0][0 : boundary_range + 1]
-        east_dec = pixel_boundaries[1][0 : boundary_range + 1]
-
-        west_ra = pixel_boundaries[0][end - boundary_range : end]
-        west_dec = pixel_boundaries[1][end - boundary_range : end]
-
-        bound_ra = np.concatenate((east_ra, west_ra), axis=None)
-        bound_dec = np.concatenate((east_dec, west_dec), axis=None)
-        polar_boundaries = np.array([bound_ra, bound_dec])
+    num_points = len(r_asc)
+    if num_points != len(dec):
+        raise ValueError(
+            f"length of r_asc ({num_points}) is different \
+                from length of dec ({len(dec)})"
+        )
+
+    if num_points == 0:
+        return np.array([])
+
+    bounds = hp.vec2dir(
+        hp.boundaries(2**pixel_order, pixel, step=step, nest=True), lonlat=True
+    )
+    margin_check = np.full((num_points,), False)
+
+    distances = _get_boundary_point_distances(pixel_order, pixel, step)
+    margin_threshold_deg = margin_threshold / 3600.
+
+    num_bounds = step * 4
+    ra_chunks = np.array_split(r_asc, chunk_size)
+    dec_chunks = np.array_split(dec, chunk_size)
+    index = 0
+    for chunk_index, ra_chunk in enumerate(ra_chunks):
+        dec_chunk = dec_chunks[chunk_index]
+
+        chunk_len = len(ra_chunk)
+
+        if chunk_len > 0:
+            ra_matrix = np.repeat(np.reshape([ra_chunk], (-1, 1)), num_bounds, axis=1)
+            dec_matrix = np.repeat(np.reshape([dec_chunk], (-1, 1)), num_bounds, axis=1)
+
+            bounds_ra_matrix = np.repeat(np.array([bounds[0]]), len(ra_chunk), axis=0)
+            bounds_dec_matrix = np.repeat(np.array([bounds[1]]), len(dec_chunk), axis=0)
+
+            points_coords = SkyCoord(ra=ra_matrix, dec=dec_matrix, unit="deg")
+            bounds_coords = SkyCoord(
+                ra=bounds_ra_matrix, dec=bounds_dec_matrix, unit="deg"
+            )
+
+            separations = points_coords.separation(bounds_coords)
```

Comment:
that's exactly it how it works, and thanks! and it could definitely be made faster with a GPU if you're volunteering to write the cuda implementing 😭 😭

Reply:
Haha, if only we had the time